### DISCUSSION TITLE
The implementation of the readEbs function in the objectnode module is not reasonable, and the buffer size is too small.

### DISCUSSION THREAD

**Discussion by FourSpaces** [cubefs/cubefs#3422] [Q&A] [answered]
The implementation of the readEbs function in the objectnode module is not reasonable. Each Buf makes a request, and the Buf size is too small, which can cause the objectnode to download Ebs data slowly.

[tmp buffer 32k ](https://github.com/cubefs/cubefs/blob/cef58ab3db04857b05a69d9a132e37d4d92e79c7/util/buf/buffer_pool.go#L18)
```
var ReadBufPool = sync.Pool{
	New: func() interface{} {
		b := make([]byte, 32*1024)
		return b
	},
}
```

[readEbs()](https://github.com/cubefs/cubefs/blob/cef58ab3db04857b05a69d9a132e37d4d92e79c7/objectnode/fs_volume.go#L1551)
```
	upper := size + offset
	if upper > inodeSize {
		upper = inodeSize - offset
	}

	ctx := context.Background()
	// _ = context.WithValue(ctx, "objectnode", 1) // TODO: ???
	reader := v.getEbsReader(inode)
	var n int
	var rest uint64
	tmp := buf.ReadBufPool.Get().([]byte)
......
```

I think the slow download speed is due to the small Buf. I adjusted the Buf size to 512K and the speed immediately improved

When the tmp Buf is 32K:
Download speed
![image](https://github.com/cubefs/cubefs/assets/13359726/09178a73-e235-42c8-8916-f6e737ac09ee)

Upload speed
![image](https://github.com/cubefs/cubefs/assets/13359726/4da94b04-3b6a-41b3-aa36-c53a77d49630)

When the tmp Buf is 512K:
Download speed
![image](https://github.com/cubefs/cubefs/assets/13359726/fb1992b5-2797-4c3e-a4bf-2feb4f4614ca)

Upload speed
![image](https://github.com/cubefs/cubefs/assets/13359726/662d313a-5a6f-4024-ab6c-c0e6b5e84ef2)

The reason why the download speed has significantly improved here and the upload speed has not increased too much is that I guess it is due to merging multiple Buf into one block size during uploading and interacting with EBS.
When downloading, interacting directly with EBS with a single Buf results in many requests, which is why the speed is slow. By increasing the Buf size, the number of requests decreases, the request overhead decreases, and the speed improves.

**Comment by tangdeyi** [comment, marked as answer]
Thank you for your feedback. The performance of reading cold volume large objects can be optimized here. In the large object download scenario, a small buff will result in a large number of blobstore access requests, resulting in a decrease in download performance. We will optimize this in the next version.
